### DISCUSSION TITLE
Authorization error

### DISCUSSION THREAD

**Discussion by mahitha-ada** [nik-base/auto-tfs#2] [General]
Hi Nikhil,

I am getting this error TF command 'checkout' error: TF30063: You are not authorized to access.

Where can I give my PAT token?

Looking forward for your reply.

**Comment by nik-base** [comment]
Hi Sandeep,

You are not required a PAT token. Are you able to checkout from Visual Studio itself as well as command line?

Also, Can you check if you can access TFS from the browser inside the visual studio. You will find the browser in View->Other Windows->Web Browser. Use your TFS url in navigation bar to access the TFS. If you get unauthorized error here, then try login to TFS here. Also, make sure you have appropriate rights for the item you are trying to checkout. If the issue persists, try restarting system/visual studio. 

Let me know if it helps.

**Comment by mahitha-ada** [comment]
for some reason, I am able to navigate to TFS url in regular chrome or FF browser but not from the visual studio->Other Windows-> Web Browser :(

**Comment by nik-base** [comment]
You are not able to navigate to TFS url or getting some error? You can login from here if it possible to transverse to login page. Or you can try removing the connection and adding it again. Also, try clearing the TFS credentials in credential manager and then again try  with the visual studio web browser.

**Comment by mahitha-ada** [comment]
Just getting that error will try clearing credential manager tfs details and try logging again

**Comment by mahitha-ada** [comment]
![image](https://user-images.githubusercontent.com/410096/130088718-7f8497e9-041f-4d42-866f-584266fe03dc.png)

**Comment by nik-base** [comment]
The extension relies on the TF command. You can try executing the TF command manually by traversing to the path where TF.exe is then execute any TF command. For example - TF status <Local Path of any mapped workspace/folder/item. If this command prompts for any credentials, try providing them and then try again.

**Comment by mahitha-ada** [comment]
I tried that but every time I had to set like this in powershell
PS C:\> Set-Alias tfs "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\TF.exe"

and after that for every file status it's asking me credentials I know it's more related to TFS

**Comment by nik-base** [comment]
Here are some links which have some solution for same issue -

https://www.codewrecks.com/post/old/2014/07/your-tfs-keeps-asking-you-for-credentials-even-with-active-directory/

https://social.msdn.microsoft.com/Forums/sqlserver/en-US/559895b1-b4d9-4267-8487-0bc6e47ad4d7/team-explorer-and-tfexe-ignoring-my-saved-usernamepassword?forum=tfsgeneral

https://social.msdn.microsoft.com/Forums/vstudio/en-US/84b4749c-2838-4975-9b61-991cb90582b4/always-prompted-for-login-credentials?forum=tfsgeneral

**Comment by nik-base** [comment]
Converting to discussion as related to TFS itself. Feel free to reopen if issue occurs event after TFS resolution

**Comment by KenBrannigan** [comment]
Should this extension still work? I am getting the same "TF command 'checkout' error: TF30063: You are not authorized to access." error. I pointing to the VS 2022 TF.exe. I have no problem from within VS 2022 performing any commands.

If I change settings to use a prompt for check-in I get a command prompt that disappears quickly but has a message that says "Unable to determine the workspace..."

**Comment by nik-base** [comment]
The extensions works with TF.exe from VS2022 as well. Are you able to run TF commands from command line? Check [TF commnds](https://docs.microsoft.com/en-us/azure/devops/repos/tfvc/use-team-foundation-version-control-commands?view=azure-devops) for different commands. Execute checkout command on the workspace that is mapped on your local.

**Reply by KenBrannigan** [reply]
When I tried via a command prompt to check in a file I got prompted with a browser window to authenticate with Azure DevOps. After providing my credentials I was provided a check-in window so that I could associate work items etc. I then was able to complete the check in process. I closed the command prompt and opened a new one and tried again and only got the window to associate work items etc. so it retained my authentication information. 

I then opened up VS code and edited a file and tried to check it in and I got the following command prompt:

![image](https://user-images.githubusercontent.com/11855023/164716203-bab9f55a-a3a0-4526-81f7-d9e50ce32e40.png)

I appreciate any help you can provide.

**Comment by nik-base** [comment]
@KenBrannigan , Have tried the command suggested in the error message on the workspace you have mapped to your local path? Also try getting latest from the workspace using command `tf get`.

**Reply by KenBrannigan** [reply]
@nik-base I ran the command from the error message and got a login prompt to provide my credentials. Once they were provided I got the expected response listing the workspaces I have access to. I also ran the tf get command and it was completed with no issues.

When I go back to VS Code and try to check in I still get the same error. 

Let me know if there is anything else I can check or try.

**Reply by KenBrannigan** [reply]
@nik-base I found the issue. I was able to grab the command that the extension is executing to perform the check-in:

"C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\TF.exe"  checkin /comment:"Testing" /validate c:\xxx\aaa bbb ccc\dev\readme.md

The path to my project has spaces in it and it is not being wrapped in double-quotes:

c:\xxx\aaa bbb ccc\dev\readme.md -> "c:\xxx\aaa bbb ccc\dev\readme.md"

Is this something you can fix for me? I assume some of the other commands that use the path would also need the same handling.

**Reply by nik-base** [reply]
@KenBrannigan , You are right, the `checkin` command does not work when there are spaces in the file path. I have fixed the issue and created `vsix` for the same (attached below). Can you confirm if it works on your end by installing the `vsix`. I will publish a new version if it works.

[auto-tfs-1.3.3.vsix.zip](https://github.com/nik-base/auto-tfs/files/8569574/auto-tfs-1.3.3.vsix.zip)

**Reply by KenBrannigan** [reply]
@nik-base that worked perfectly. I am also now able to compare files etc. Thank you very much for making the change.

One additional question. We use gated check-ins. Is it possible for the extension to give you the status of the gated check-in and once complete allows you to reconcile the workspace as Visual Studio does? If not no worries I can use the Dev Ops website to see when the check-in is complete and manually execute a tf.exe reconcile /clean /ignore.

**Reply by mahitha-ada** [reply]
This works for me too. I am glad I subscribed to the discussions. I am able to use the new Vsix file and able to successfully checkout, get latest and other stuff changed files are even shown in the source control tab. Please release the version it would be be great. Is there a way to checkin the files?

**Reply by KenBrannigan** [reply]
You can click the icon in the bottom status bar to do a check-in:

![image](https://user-images.githubusercontent.com/11855023/165640531-c1a5c105-9f9b-4b9f-80e7-b56d4492e2bb.png)

**Reply by mahitha-ada** [reply]
Thank you for prompt reply Ken, it says provide a comment for this checkin and that would be on top of sourcecontrol rite?

**Reply by KenBrannigan** [reply]
No problem @sandeesaripalli. Yes the top of the source control section is for the comments:

![image](https://user-images.githubusercontent.com/11855023/165647759-f6c498b7-4168-4c9f-b745-6c70b667220f.png)

I also change the settings so that I get a check-in prompt which gives me more options like work item association etc:

![image](https://user-images.githubusercontent.com/11855023/165647888-09fcab79-57e7-4fca-b231-001a82650fd8.png)

**Reply by nik-base** [reply]
@KenBrannigan , I did not find any command that gives status of the gated check-in. When you check-in your code does the check-in command succeeds immediately or does it complete only after passing all the gates?

**Reply by nik-base** [reply]
@KenBrannigan @sandeesaripalli , You also have option to check-in/shelve in the source control side bar.

![image](https://user-images.githubusercontent.com/35769107/165687030-a9e57250-ff15-4c93-82b2-b71f8332ac9a.png)

You have make sure source control repositories is checked in the source control context menu.

![image](https://user-images.githubusercontent.com/35769107/165687436-22e17c1e-7a17-4b0c-a188-da591da3c213.png)

**Reply by KenBrannigan** [reply]
@nik-base it looks like the check-in command completes immediately with a gated check-in and gives you a specific message that mentions a gated check-in. Looks like the only way would not be with TF.exe that I can see and instead it would have to talk directly to Dev Ops API to see the status of the check-in. At minimum can we make the reconcile command part of the extension?

**Reply by nik-base** [reply]
@KenBrannigan Sure, I can release a version with reconcile command. I have not personally used reconcile command. Can you help with what parameters would be required for this command referred [here](https://docs.microsoft.com/en-us/azure/devops/repos/tfvc/reconcile-command?view=azure-devops). Does this command need to be executed only at the root of workspace? And would the reconcile button be placed in source control tab?

**Comment by nik-base** [comment]
@KenBrannigan @sandeesaripalli , I have published a new version with the changes on the marketplace.

[Auto TFS v1.3.3](https://marketplace.visualstudio.com/items?itemName=nikhil-patil.auto-tfs)

**Reply by KenBrannigan** [reply]
@nik-base thank you for pushing this update out I really appreciate you handling this change so quickly. I will dig a bit deeper into the reconcile command that Visual Studio runs and get back to you with details about which options I would need.
